### PR TITLE
Start the definition of concrete number and container interfaces

### PR DESCRIPTION
**Pull request by ChrisRackauckas**

### CODE REVIEW

**Comment by ChrisRackauckas** [comment]
Comments @avik-pal @oscardssmith @YingboMa?

File: docs/src/interfaces/Array_and_Number.md [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+!!! note
+
+    There are many types which can work with a specific solver that do satisfy this
+    interface. Many times as part of prototyping you may want to side-step the
+    high level interface checks in order to simply test whether a new type is working.
+    To do this, set `interface_checks = false` as a keyword argument to `init`/`solve`
+    to bypass any of the internal interface checks. This means you will no longer get
+    a nice high-level error message and instead it will attempt to use the type
+    without restrictions. Note that not every problem/solver has not implemented this
+    new keyword argument as of 2023.
```
Comment (ranocha) [inline docs/src/interfaces/Array_and_Number.md:21]:
```suggestion
    without restrictions. Note that not every problem/solver has implemented this
    new keyword argument as of 2023.
```

File: docs/src/interfaces/Array_and_Number.md [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+
+* `similar(x::T)::T`
+* `zero(x::T)::T`
+* `z::T .= x::T .+ y::T` is defined
+* `z::T .= x::T .* y::T` is defined
+* `z::T .= t::T2 .* x::T` where `T2` is the time type (a Number) and `T` is the container type.
+
+### SciML Matrix (Operator) Type
+
+Note that the matrix type may not match the type of the initial container `u0`. An example is `ComponetMatrix`
```
Comment (ranocha) [inline docs/src/interfaces/Array_and_Number.md:94]:
```suggestion
Note that the matrix type may not match the type of the initial container `u0`. An example is `ComponentMatrix`
```

File: docs/src/interfaces/Array_and_Number.md [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+
+Additionally, the following rules apply to subsets of uses:
+
+### SciML Mutable Array Types
+
+* `similar(x::T)::T`
+* `zero(x::T)::T`
+* `z::T .= x::T .+ y::T` is defined
+* `z::T .= x::T .* y::T` is defined
+* `z::T .= t::T2 .* x::T` where `T2` is the time type (a Number) and `T` is the container type.
```
Comment (ChrisRackauckas) [inline docs/src/interfaces/Array_and_Number.md:90]:
```suggestion
* `z::T .= t::T2 .* x::T` where `T2` is the time type (a Number) and `T` is the container type.
* (Optional) `Base.resize!(x,i)` is required for `resize!(integrator,i)` to be supported.
```

File: docs/pages.jl [hunk 4]
```diff
@@ -2,8 +2,11 @@
 
 pages = [
     "Home" => "index.md",
-    "Interfaces" => Any["interfaces/Problems.md",
+    "Interfaces" => Any[
+        "interfaces/Array_and_Number.md",
+        "interfaces/Problems.md",
         "interfaces/SciMLFunctions.md",
+        "interfaces/Array_and_Number.md",
```
Comment (ChrisRackauckas) [inline docs/pages.jl:9]:
```suggestion
```

File: docs/src/interfaces/Array_and_Number.md [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+* `one(x::T) * oneunit(x::T) = z::T`
+* `t::T2 * x::T + y::T = z::T` for `T2` a time type and `T` the dependent variable type (this includes the
+  `muladd` equivalent form).
+
+Additionally, the following rules apply to subsets of uses:
+
+### Adaptive Number Types
+
+* `x::T / y::T = z::T`
+* Default choices of norms can assume `sqrt(x::T)::T` exists. If `internalnorm` is overriden then this
```
Comment (ranocha) [inline docs/src/interfaces/Array_and_Number.md:52]:
```suggestion
* Default choices of norms can assume `sqrt(x::T)::T` exists. If `internalnorm` is overridden then this
```

**Pull request merged**